contracts/RelayerHub.sol

```diff
@@ -8,88 +8,206 @@ import "./System.sol";
-  }
+contract RelayerHub is IRelayerHub, System, IParamSubscriber {
+    using SafeMath for uint256;
+
+    bool public alreadyUpdate;
+
+    uint256 public constant INIT_REQUIRED_DEPOSIT = 1e20;
+    uint256 public constant INIT_DUES = 1e17;
+    address public constant WHITELIST_1 = 0xb005741528b86F5952469d80A8614591E3c5B632;
+    address public constant WHITELIST_2 = 0x446AA6E0DC65690403dF3F127750da1322941F3e;
+
+    uint256 public requiredDeposit;
+    uint256 public dues;
+
+    mapping(address => relayer) relayers; // old map holding the relayers which are to be allowed safe exit
+    mapping(address => bool) relayersExistMap;
+
+    mapping(address => manager) managers;
+    mapping(address => bool) managersRegistered;
+    mapping(address => bool) relayManagersExistMap;
+    mapping(address => address) managerToRelayer;
+    mapping(address => bool) currentRelayers;
+
+    struct manager {
+        uint256 deposit;
+        uint256 dues;
+    }
+
+    modifier onlyPreviousRegisteredRelayer() {
+        require(relayersExistMap[msg.sender], "relayer does not exist");
+        _;
+    }
+
+    modifier onlyNonRegisteredManager() {
+        require(relayManagersExistMap[msg.sender], "manager does not exist");
+        _;
+    }
+
+    modifier onlyRegisteredManager() {
+        require(relayManagersExistMap[msg.sender], "manager does not exist");
+        require(managersRegistered[msg.sender], "manager not registered");
+        _;
+    }
+
+    modifier noProxy() {
+        require(msg.sender == tx.origin, "no proxy is allowed");
+        _;
+    }
+
+    modifier onlyAllowedParty() {
+        require(msg.sender == 0xb005741528b86F5952469d80A8614591E3c5B632 || msg.sender == 0x446AA6E0DC65690403dF3F127750da1322941F3e, "the msg sender is not allowed to call update to ensure smooth transition");
+        // todo change the above address to appropriate ones which can call update()
+        _;
+    }
+
+    event relayerRegister(address _relayer);
+    event relayerUnRegister(address _relayer);
+    event paramChange(string key, bytes value);
+
+    event removeManagerByGovEvent(address _removedManager);
+    event addManagerByGovEvent(address _addedManager);
+    event registerManagerEvent(address _registeredManager);
+    event updateRelayerEvent(address _from, address _to);
+
+    function exitOldRelayers() external onlyPreviousRegisteredRelayer {
+        relayer memory r = relayers[msg.sender];
+        msg.sender.transfer(r.deposit.sub(r.dues));
+        address payable systemPayable = address(uint160(SYSTEM_REWARD_ADDR));
+        systemPayable.transfer(r.dues);
+        delete relayersExistMap[msg.sender];
+        delete relayers[msg.sender];
+        emit relayerUnRegister(msg.sender);
+    }
+
+    function init() external onlyNotInit {
+        requiredDeposit = INIT_REQUIRED_DEPOSIT;
+        dues = INIT_DUES;
+        alreadyInit = true;
+    }
+
+    function update() external onlyAllowedParty {
```

Comment:
The `update` name is too generic, and I think `onlyAllowedParty` is not necessary, let us allow anyone to invoke it, there is no risk.